app/uk/gov/hmrc/iossreturns/services/FinancialDataService.scala

```diff
@@ -17,23 +17,53 @@
-import uk.gov.hmrc.iossreturns.connectors.FinancialDataHttpParser.FinancialDataResponse
 import uk.gov.hmrc.iossreturns.models.Period
-import uk.gov.hmrc.iossreturns.models.financialdata.FinancialDataQueryParameters
+import uk.gov.hmrc.iossreturns.models.financialdata.{FinancialData, FinancialDataQueryParameters}
 import uk.gov.hmrc.iossreturns.models.payments.Charge
 
 import java.time.LocalDate
 import javax.inject.Inject
 import scala.concurrent.{ExecutionContext, Future}
 
-class FinancialDataService @Inject()(financialDataConnector: FinancialDataConnector)(implicit executionContext: ExecutionContext) {
-  def getFinancialData(iossNumber: String, fromDate: Option[LocalDate], toDate: Option[LocalDate]): Future[FinancialDataResponse] = {
-    financialDataConnector.getFinancialData(iossNumber, FinancialDataQueryParameters(fromDate = fromDate, toDate = toDate))
-  }
+class FinancialDataService @Inject()(financialDataConnector: FinancialDataConnector, returnsService: ReturnsService)(implicit executionContext: ExecutionContext) {
 
   def getCharge(iossNumber: String, period: Period): Future[Option[Charge]] = {
-    getFinancialData(iossNumber, Some(period.firstDay), Some(period.lastDay)).map { maybeFinancialData =>
+    getFinancialDataForDateRange(iossNumber, period.firstDay, period.lastDay).map { maybeFinancialData =>
       maybeFinancialData.flatMap(_.getChargeForPeriod(period))
     }
   }
+
+  def getFinancialData(iossNumber: String, fromDate: Option[LocalDate], toDate: Option[LocalDate]): Future[Option[FinancialData]] = {
+    (fromDate, toDate) match {
+      case (Some(dateFrom), Some(dateTo)) =>
+        val financialDatas: Future[Seq[FinancialData]] =
+          Future.sequence(
+            returnsService.getPeriodYears(dateFrom, dateTo).map {
+              taxYear => getFinancialDataForDateRange(iossNumber, taxYear.startOfYear, taxYear.endOfYear)
+            }
+          ).map(_.flatten)
+
+        financialDatas.map {
+          case firstFinancialData :: Nil => Some(firstFinancialData)
+          case firstFinancialData :: rest =>
+            val otherFinancialTransactions = rest.flatMap(_.financialTransactions).flatten
+
+            val allTransactions =
+              firstFinancialData.financialTransactions.getOrElse(Nil) ++ otherFinancialTransactions
+
+            val maybeAllTransactions = if (allTransactions.isEmpty) None else Some(allTransactions)
+            Some(firstFinancialData.copy(financialTransactions = maybeAllTransactions))
+          case Nil => None
+        }
+      case _ =>
+        Future.failed(new IllegalArgumentException("Both fromDate and toDate must be provided"))
+    }
+  }
+
+  private def getFinancialDataForDateRange(iossNumber: String, fromDate: LocalDate, toDate: LocalDate): Future[Option[FinancialData]] = {
+    financialDataConnector.getFinancialData(iossNumber, FinancialDataQueryParameters(fromDate = Some(fromDate), toDate = Some(toDate))).flatMap {
```

Comment:
`.flatMap` not needed here, just a `.map` should suffice and you can remove the `Future.successful`

Also, I think the indentation isn't right for the case's